lib/Core/SpecialFunctionHandler.cpp

```diff
@@ -1023,48 +1086,36 @@ void SpecialFunctionHandler::handleMarkPersistent(ExecutionState &state,
   }
 
-  Executor::ExactResolutionList rl;
   for (uint64_t offset = 0; offset < realSize; offset += PersistentState::MaxSize) {
-    ref<Expr> offsetExpr = ConstantExpr::create(offset, Expr::Int64);
-    ref<Expr> ptrExpr = AddExpr::create(addr, offsetExpr);
-    executor.resolveExact(state, ptrExpr, rl, "mark_persistent");
-  }
-  
-  for (Executor::ExactResolutionList::iterator it = rl.begin(), 
-         ie = rl.end(); it != ie; ++it) {
-    const MemoryObject *mo = it->first.first;
+    bool succ;
+    ObjectPair found;
+    ref<Expr> offsetExpr = ConstantExpr::create(offset, addr->getWidth());
+    ref<Expr> currAddr = AddExpr::create(addr, offsetExpr);
+    assert(state.addressSpace.resolveOne(state, executor.solver, currAddr, found, succ));
+    assert(succ);
+    const MemoryObject *mo = found.first;
     mo->setName(name);
-    // klee_warning("Marking object at %p as persistent", (void*)mo->address);
-    
-    ExecutionState *s = it->second;
+
     // FIXME: Type coercion should be done consistently somewhere.
+    // Check that the size is proper.
+    uint64_t properSize = std::min(PersistentState::MaxSize, realSize - offset);
+    ref<Expr> properSizeExpr = ConstantExpr::create(properSize, offsetExpr->getWidth());
+    ref<Expr> sizeCheckExpr = EqExpr::create(properSizeExpr, mo->getSizeExpr());
     bool res;
-    // bool success __attribute__ ((unused)) =
-    //   executor.solver->mustBeTrue(*s, 
-    //                               EqExpr::create(ZExtExpr::create(arguments[1],
-    //                                                               Context::get().getPointerWidth()),
-    //                                              mo->getSizeExpr()),
-    //                               res);
-    bool success __attribute__ ((unused)) =
-      executor.solver->mustBeTrue(*s, 
-                                  EqExpr::create(ConstantExpr::create(PersistentState::MaxSize,
-                                                                  Context::get().getPointerWidth()),
-                                                 mo->getSizeExpr()),
-                                  res);
+    bool success = executor.solver->mustBeTrue(state, sizeCheckExpr, res);
     assert(success && "FIXME: Unhandled solver failure");
     
     if (res) {
-      executor.executeMarkPersistent(*s, mo);
-      // Just return the same address as what we received.
-      // -- This may get re-executed, but technically "s" can be a different state.
-      executor.bindLocal(target, *s, addr);
-    } else {      
-      executor.terminateStateOnError(*s, 
+      executor.executeMarkPersistent(state, mo);
+    } else {
+      executor.terminateStateOnError(state, 
                                      "wrong size given to klee_pmem_mark_persistent[_name]", 
                                      Executor::User);
     }
   }
 
+  // Just return the same address as what we received.
+  executor.bindLocal(target, state, addr);
```

Comment:
Why do we even have to bindLocal? What's this doing?

Reply:
Return value. klee_pmem_mark_persistent returns it's argument.

Reply:
Oh oh oh I understand now, totally slipped my mind that you can't just return a value from a SpecialFunction handler and have it get back to the interpreter